### PR TITLE
Add SWIFT_STDLIB_HAS_ENVIRON to remove usage of getenv/environ from stdlib

### PR DESCRIPTION
**Pull request by kubamracek**

### CODE REVIEW

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift [hunk 1]
```diff
@@ -213,6 +213,21 @@ func posixPipe() -> (readFD: CInt, writeFD: CInt) {
...
+@_silgen_name("_NSGetEnviron")
+func _NSGetEnviron() -> UnsafeMutablePointer<UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>>
+
+var environ: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?> {
+  #if os(macOS)
+  return _NSGetEnviron().pointee
+  #elseif os(Linux)
+  return __environ
+  #else
+  #error("unsupported platform")
```
Comment (compnerd) [inline stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:226]: Please add Windows support.  Windows uses `__environ` for this, but needs to be imported via the module as there is dll storage to be associated with it.

File: stdlib/public/SwiftShims/LibcOverlayShims.h [hunk 2]
```diff
@@ -63,6 +64,7 @@ static inline char * _Nullable *_swift_stdlib_getEnviron() {
   return *_NSGetEnviron();
 }
 #endif
```
Comment (compnerd) [inline stdlib/public/SwiftShims/LibcOverlayShims.h:66]: Interesting, I'm not sure why Windows is missing here...

**Review by compnerd — CHANGES_REQUESTED** [review]
This is actually a very exciting change for windows - UAP doesn't allow environment, so this is definitely a useful thing.

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift (hunk 1 above)
Comment (kubamracek) [reply stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:226]: Which module? The file already has `import CRT`, is that all that's needed?

**Comment by kubamracek** [comment]
@swift-ci please test

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift (hunk 1 above)
Comment (kubamracek) [reply stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:226]: And I can't really test how this behaves on Windows with SWIFT_STDLIB_HAS_ENVIRON=0... Can I rely on you to sign off on the support for Windows here?

**Comment by kubamracek** [comment]
@swift-ci please test Linux platform

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test Linux platform

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift (hunk 1 above)
Comment (compnerd) [reply stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:226]: Sure, that would be fine.  Yes, `CRT` should provide `environ`.

**Comment by kubamracek** [comment]
@swift-ci please test

File: stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift [hunk 3]
```diff
@@ -213,6 +213,21 @@ func posixPipe() -> (readFD: CInt, writeFD: CInt) {
...
+#if !SWIFT_STDLIB_HAS_ENVIRON
+@_silgen_name("_NSGetEnviron")
+func _NSGetEnviron() -> UnsafeMutablePointer<UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>>
+
+var environ: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?> {
+  #if os(macOS)
+  return _NSGetEnviron().pointee
+  #elseif os(Linux) || os(Windows)
+  return __environ
+  #else
```
Comment (compnerd) [inline stdlib/private/SwiftPrivateLibcExtras/Subprocess.swift:225]: Sorry, I think that we got something confused here ...

```suggestion
  #elseif os(Windows)
  return __p_environ().pointee
  #elseif os(Linux)
  return __environ
  #else
```

**Review by compnerd — CHANGES_REQUESTED** [review]

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Review by compnerd — APPROVED** [review]

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Pull request merged**